agent/cache/cache.go

```diff
@@ -555,46 +555,41 @@ func makeEntryKey(t, dc, peerName, token, key string) string {
 	return fmt.Sprintf("%s/%s/%s/%s", t, dc, token, key)
 }
 
-// fetch triggers a new background fetch for the given Request. If a
-// background fetch is already running for a matching Request, the waiter
-// channel for that request is returned. The effect of this is that there
-// is only ever one blocking query for any matching requests.
-//
-// If allowNew is true then the fetch should create the cache entry
-// if it doesn't exist. If this is false, then fetch will do nothing
-// if the entry doesn't exist. This latter case is to support refreshing.
-func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ignoreExisting bool) <-chan struct{} {
-	// We acquire a write lock because we may have to set Fetching to true.
+// fetch triggers a new background fetch for the given Request. If a background
+// fetch is already running or a goroutine to manage that still exists for a
+// matching Request, the waiter channel for that request is returned. The
+// effect of this is that there is only ever one blocking query and goroutine
+// for any matching requests.
+func (c *Cache) fetch(key string, r getOptions) <-chan struct{} {
+	// We acquire a write lock because we may have to set HasGoroutine or
+	// Fetching to true.
 	c.entriesLock.Lock()
 	defer c.entriesLock.Unlock()
+
 	ok, entryValid, entry := c.getEntryLocked(r.TypeEntry, key, r.Info)
 
-	// This handles the case where a fetch succeeded after checking for its existence in
-	// getWithIndex. This ensures that we don't miss updates.
-	if ok && entryValid && !ignoreExisting {
+	switch {
+	case ok && entryValid:
+		// This handles the case where a fetch succeeded after checking for its
+		// existence in getWithIndex. This ensures that we don't miss updates.
 		ch := make(chan struct{})
 		close(ch)
 		return ch
-	}
 
-	// If we aren't allowing new values and we don't have an existing value,
-	// return immediately. We return an immediately-closed channel so nothing
-	// blocks.
-	if !ok && !allowNew {
-		ch := make(chan struct{})
-		close(ch)
-		return ch
-	}
+	case ok && entry.Fetching:
+		// If we already have an entry and it is actively fetching, then return
+		// the currently active waiter.
+		return entry.Waiter
 
-	// If we already have an entry and it is actively fetching, then return
-	// the currently active waiter.
-	if ok && entry.Fetching {
+	case ok && entry.HasGoroutine:
```

Comment:
New conditional branch to handle the "not fetching, but there's already a goroutine" scenario underlying the bug.